src/robot.coffee

```diff
@@ -184,24 +190,112 @@ class Robot
       ((msg) -> msg.message = msg.message.message; callback msg)
     )
 
+  # Public: Registers new middleware for execution after matching but before
+  # Listener callbacks
+  #
+  # middleware - A function that determines whether or not a given matching
+  #              Listener should be executed. The function is called with
+  #              (robot, listener, response, next, done). If execution should
+  #              continue (next middleware, Listener callback), the middleware
+  #              should call the 'next' function with 'done' as an argument.
+  #              If not, the middleware should call the 'done' function with
+  #              no arguments.
+  #
+  # Returns nothing.
+  listenerMiddleware: (middleware) ->
+    if middleware.length != 5
+      throw new Error("Incorrect number of arguments for middleware callback (expected 5, got #{middleware.length})")
+    @middleware.listener.push (robot, context, next, done) ->
+      middleware.call(undefined, robot, context.listener, context.response, next, done)
+    return undefined
+
+  # Execute all middleware in order and call 'next' with the latest 'done'
+  # callback if last middleware calls through. If all middleware is compliant,
+  # 'done' should be called with no arguments when the entire round trip is
+  # complete.
+  #
+  # whichMiddleware - String name of middleware set to execute (currently
+  #     just 'listener')
+  #
+  # context - context object that is passed through the middleware stack.
+  #     When handling errors, this is assumed to have a `response` property.
+  #
+  # next(context, done) - Called when all middleware is complete (assuming
+  #     all continued by calling respective 'next' functions)
+  #
+  # done() - Initial (final) completion callback. May be wrapped by
+  #     executed middleware.
+  #
+  # Returns nothing
+  # Returns before executing any middleware
+  executeMiddleware: (whichMiddleware, context, next, done) ->
+    if not @middleware.hasOwnProperty whichMiddleware
+      throw new Error "Invalid middleware set: \"#{whichMiddleware}\""
+
+    allMiddleware = @middleware[whichMiddleware]
+
+    # Execute a single piece of middleware and update the completion callback
+    # (each piece of middleware can wrap the 'done' callback with additional
+    # logic).
+    executeMiddleware = (doneFunc, middlewareFunc, cb) =>
```

Comment:
Having this variable `executeMiddleware` inside a function `executeMiddleware` is a little confusing.